### ISSUE TITLE
[SPIKE] Decide expected types of cluster installation

### ISSUE THREAD

**Issue by mkyc** [epiphany-platform/epiphany#1819]
**Describe the solution you'd like**
Decide how to proceed with developing self-managed kubernetes services module.

1.  standard (1 * master + x * nodes)
2.  single node
3. ha (x * masters + y * nodes)

For now we agreed, that first scenario is the most important so It has to be done first. For details please refer to comments.

**Additional context**
 The goal of this issue is: create a new issues with well described steps to create such module.

**Comment by mkyc** [comment]
In this task we should identify all required scenarios for this module to perform. My initial thinking would be that we want to have separate modules for:
 - standard (1 * master + x * nodes)
 - single node
 - ha (x * masters + y * nodes)
I would focus just on standard scenario in this module.

**Comment by rafzei** [comment]
Currently, modules are prepared for the 'standard' scenario. @toszo do we have to take care of single node and ha options as well?

**Comment by toszo** [comment]
Yes, a single node is important because of the large number of installations in prod. 
HA is less popular but is also a must-have.

**Comment by mkyc** [comment]
I believe that it's impossible to finish all scenarios in reasonable time. Can agree on order how we want to work on that? Is one node more important than standard? What you want to see as result of first task? Master + 3 nodes, or single-node? 

My opinion is master + 3 nodes is most standard think we want. I'd like to have it as first version.

**Comment by toszo** [comment]
Hmm but currently we support all 3 scenarios with epicli, are we going to drop some functionality?

**Comment by mkyc** [comment]
Not drop, but order. 

I believe we should start with "1master + X nodes" scenario in this module. During development of this scenario we will be able to see if single-node scenario is simple to add with some parameter or not. Then we will be able to decide if we want to move this single-node scenario into separate module. The same with HA: we will be able to judge during development if it's easy to introduce HA with parameter or maybe we want to exclude it to another module. 

Let's remember, that simple but rock-solid module is better than multipurpose but poor one. I believe that we can have 1-3 modules as result of this epic, not just 1 module. 

Thing is only to agree on order of work. Roughly estimating it's 4 workable weeks till the end of this year, and I do not know if we can do all scenarios at once.

**Comment by toszo** [comment]
Ok, since we will still have the possibility to run epicli in "old fashioned" way I'm fine with go with it in this module.

**Comment by rafzei** [comment]
I see we have agreement here so I will push it to to-do. 
The goal of this issue is: create a new issues with well described steps to create such module.
(Consider as reference: Epic K8s Autoscaler)

**Comment by mkyc** [comment]
@rafzei please update description of that issue as well.

**Comment by sk4zuzu** [comment]
All the k8s installation engines I have checked in the Spike https://github.com/epiphany-platform/epiphany/pull/1893 are HA by default. There should be no obvious distinction between `1:N` or `M:N` or even `1:1`, because all the tools likely support all the scenarios.
I wondered from where this confusion came from and maybe this is because in Epiphany for historical reasons we have to set special flags in shared config to "enable" ha mode. This is not the case with `rke` or `kubespray` in fact it's easier and safer to implement just the ha scenario in such tools and allow setting `1:1` and others. :thinking:

**Comment by atsikham** [comment]
@sk4zuzu @toszo @mkyc @rafzei Does anyone think that at the moment there should be other tasks created than #1912? The goal there is to create initial RKE module for cluster installation and as a result also  to create tasks for improvements.

**Comment by erzetpe** [comment]
@atsikham For this moment I think that we can have additionally to [#1912](https://github.com/epiphany-platform/epiphany/issues/1912) task related to check if/how much of the code from existing solution we can move into the module.

I think no matter if we will port our existing code for Kubernetes or we will start with the new development split into three parts the result should be as described by sk4zuzu - so there should be no distinction for user if he wants to have single node, single master or HA setup done except the configuration of the module (e.g. number of masters, availability to run other than system pods on master).

**Comment by seriva** [comment]
Closing as work on modules is halted and this is no longer relevant. If needed we can revisit.
